fix(bigquery): Make exponential backoff retry second based

mina-asham opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigquerystorage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #2211 ☕️

- Currently these retries are doing retries in milliseconds, which is way too aggressive, the current retry steps are (1ms, 2ms, 4ms, 8ms, 16ms, 32ms, 64ms, 128ms, 256ms, 512ms, 1024ms, ...), it's highly likely the first 7 retries will fail with rate limiting specially in a big workload (specifically this is causing an issue in the Spark bigquery connector), I suspect the exponential retries here were meant to be in seconds (though even if it wasn't meant to, this might be a slightly better approach to reduce the load on the servers
- The new steps would be: 1s, 2s, 4s, 8s, 16s, 32s, 60s (repeated till we exhaust the 5 minutes max retries)

mina-asham committed:
fix(bigquery): Make exponential backoff retry second based

- Currently these retries are doing retries in milliseconds, which is way too aggressive, the current retry steps are (1ms, 2ms, 4ms, 8ms, 16ms, 32ms, 64ms, 128ms, 256ms, 512ms, 1024ms, ...), it's highly likely the first 7 retries will fail with rate limiting specially in a big workload (specifically this is causing an issue in the Spark bigquery connector), I suspect the exponential retries here were meant to be in seconds (though even if it wasn't meant to, this might be a slightly better approach to reduce the load on the servers
- The new steps would be: 1s, 2s, 4s, 8s, 16s, 32s, 60s (repeated till we exhaust the 5 minutes max retries)

GaoleMeng commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:
```diff
@@ -502,7 +502,7 @@ private void maybeWaitForInflightQuota() {
   @VisibleForTesting
   static long calculateSleepTimeMilli(long retryCount) {
-    return Math.min((long) Math.pow(2, retryCount), 60000);
+    return Math.min((long) Math.pow(2, retryCount) * 1000, 60000);
```
Thanks for submitting this, the original intention for retrying reconnection at so frequent level is the majority of the connection reconnection does not need seconds of waiting. Were you seeing retry failure with small retry interval?

mina-asham replied:
The problem I saw is probably more throttling rather than specific failures, this is used directly in the bigquery-spark-connector and the problem I am seeing is the connector slowing down to a failure writing data into bigquery, this is one of the two points I found led to this (there might be more, I might be mistaken too), Seconds maybe too much yes, bit retrying within a few millis is too aggressive I think, how about we do a base of 50ms or 100ms instead of 1s? That would get the best of both worlds, retry fast the first few times, and then back off exponentially

GaoleMeng replied:
Yeah, I think 50ms makes sense to me. Let's use that value

mina-asham replied:
Updated to 50ms

mina-asham committed:
Update retry base to 50ms

mina-asham committed:
Update tests

mina-asham commented:
@shollyman can we merge this one?

mina-asham commented:
@GaoleMeng what's the next steps here, this has been approved for almost 3 months now?

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

mina-asham committed:
Avoid long overflow

- Make the comparison happen on the double side then cast to long at the end to avoid overflow happening from the calculation itself
- This was actually a bug from before, just wasn't happening because there was no 50 multiplier to get it into that state!
